[homeconnect] Home Connect Binding initial contribution

bruestel opened the pull request:
**Home Connect Binding**

The binding integrates the Home Connect system into openHAB. It uses the Home Connect API to connect to household devices (mainly Bosch and Siemens).

- Home Connect homepage: [https://www.home-connect.com](https://www.home-connect.com)
- Home Connect developer site: [https://developer.home-connect.com](https://developer.home-connect.com)

A snapshot of the binding can be downloaded from the community page [Home Connect addon: Looking for beta testers](https://community.openhab.org/t/home-connect-addon-looking-for-beta-testers/49702).

mstroeve commented:
Maybe its good to mention this binding was already reviewed multiple times :-). See #6075 
Also #3447 can be closed in my opinion.

andrewfg commented:
Hi @bruestel I have been using your your latest build, and I am still getting huge warning messages in the openhab.log as shown below.

These are errors due to the Bosch remote server being down (or something like that), they are not caused by your binding, and they do eventually correct themselves; but in my case they seem to occur very frequently, and they fill up the log file with a huge amount of "horror" warnings. I think this could shock and confuse the average user. So, therefore, I suggest you split your logging into two levels as follows, so that the average user will only see a short message..

1. logger.warn("Bosch remote server: something happened! (user tip: log set trace for more info)");
2. logger.trace(" ... the full multi line dump .. ");

```
2020-02-12 23:03:08.875 [WARN ] [internal.client.HomeConnectApiClient] - [API_ERROR BOSCH-CMG676BS6B-*********] GET 409 https://api.home-connect.com/api/homeappliances/BOSCH-CMG676BS6B-*******/status/BSH.Common.Status.RemoteControlStartAllowed
> authorization: [Bearer ******* ]
> accept: [application/vnd.bsh.sdk.v1+json]

< date: [Wed, 12 Feb 2020 23:03:08 GMT]
< content-length: [209]
< vary: [Accept-Encoding]
< access-control-allow-headers: [Authorization,Content-Type,Accept,Cache-Control,Last-Event-ID,If-Modified-Since,X-Requested-With]
< x-frame-options: [SAMEORIGIN]
< access-control-allow-methods: [GET,POST,OPTIONS,PUT,DELETE]
< access-control-allow-origin: [*]
< access-control-allow-credentials: [true]
< x-content-type-options: [nosniff]
< x-xss-protection: [1; mode=block]
< referrer-policy: [origin]
< connection: [close]
< content-type: [application/vnd.bsh.sdk.v1+json]
< hc-env: [EU-PRD]
{
  "error": {
    "key": "SDK.Error.HomeAppliance.Connection.Initialization.Failed",
    "description": "HomeAppliance did not respond to connection initialization requests in time, it might be offline"
  }
}
```

andrewfg commented:
I have been using this binding in my operative system for more than three months. It turns out that my earlier comment concerning the "huge messages" was due to a server side error at Bosh Siemens Hausgeraete, and since February it has been full stable, so IMHO this binding is fully ready to be officially reviewed  and adopted into openHAB.

mhilbush commented on bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/type/HomeConnectDynamicStateDescriptionProvider.java:
```diff
@@ -0,0 +1,73 @@
+            return stateDescriptions.get(channel.getUID().getAsString());
+        }
+
+        return originalStateDescription;
```
You should not return the originalStateDescription here. It will break other bindings that have dynamic state descriptions.

See [here](https://community.openhab.org/t/dynamic-state-options-not-showing-up/95625) and [here](https://community.openhab.org/t/squeezebox-favorites/39355/69?u=mhilbush) for more information.

Return null instead.

bruestel replied:
Thanks for the information. I will change it.

martinvw replied:
And another one bites the dust: https://community.openhab.org/t/heos-denon-support/3449/322?u=martinvw

martinvw commented on bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/servlet/AbstractServlet.java:
```diff
@@ -0,0 +1,83 @@
+
+    protected String replaceKeysFromMap(String template, Map<String, String> map) {
+        final Matcher m = MESSAGE_KEY_PATTERN.matcher(template);
+        final StringBuffer sb = new StringBuffer();
```
FYI, normally you should use the `StringBuilder` its faster and not threadsafe but that should be okay for a local variable.

PS: this is not a full review

martinvw commented on bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/client/HomeConnectSseClient.java:
```diff
@@ -0,0 +1,240 @@
+                    }
+
+                    if (message != null && !StringUtils.isEmpty(message) && !EMPTY_EVENT.equals(message)) {
+                        ArrayList<Event> events = mapToEvents(message, haId);
```
```suggestion
                        mapToEvents(message, haId).forEach(eventListener::onEvent);
```

martinvw commented on bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/client/HomeConnectSseClient.java:
```diff
@@ -0,0 +1,240 @@
+                    }
+
+                    if (message != null && !StringUtils.isEmpty(message) && !EMPTY_EVENT.equals(message)) {
+                        ArrayList<Event> events = mapToEvents(message, haId);
+                        events.forEach(e -> eventListener.onEvent(e));
```
```suggestion
                        mapToEvents(message, haId).forEach(eventListener::onEvent);
```

martinvw commented on bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/client/HomeConnectSseClient.java:
```diff
@@ -0,0 +1,240 @@
+                                    "Stopping listener! Got FORBIDDEN response from server. Please check if you allowed to access this device.");
+                            retry = false;
+                        } else if (response.code() == HTTP_UNAUTHORIZED) {
+                            logger.errorWithHaId(haId, "Stopping listener! Access token became invalid.");
```
In [E.4 of the coding guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging) we listed what kind of logging level should be used when. Error should only be used:

> `error` logging should only be used to inform the user that something is tremendously wrong in his setup, the system cannot function normally anymore, and there is a need for immediate action. 

Please change this to use the correct logging levels and if possible change the ONLINE/OFFLINE status of thing / bridge.

Alluuu replied:
Well, to be fair, if your access token becomes (or is invalid) it is a user configuration issue, that has to be changed immediately in order to continue the system functioning, without user interaction inputting valid token, at least in the closed system of Bosch home connect's sense the system is no longer functioning.

martinvw commented on bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/logger/LogWriter.java:
```diff
@@ -0,0 +1,315 @@
+ * @author Jonas Brüstel - Initial Contribution
+ */
+@NonNullByDefault
+public class LogWriter {
```
Do we really need this, it feels very heavy, what about using the MDC:

http://logback.qos.ch/manual/mdc.html

Could that make sense and do we really need all the context, I don't think bindings should contain code like this.

cpmeister replied:
Yea, I didn't want to address this in my first pass review, but I don't think that bindings should create their own logging framework. I realize that you are trying to log to have the binding log to its own file but you should leave it up to the users to configure that through the logback configuration files.

IMO all of the custom logging code should be removed and replaced with the standard slf4j logging.

bruestel commented:
> Hi @bruestel Jonas, is there any way we can move this one forward? I use the binding every day, and I would very much like to it to become adopted as an official part of OpenHAB. I am willing to (help) modifying your code to address the issues raised by the reviewers. But I would only do it if you are willing to accept me so doing...

Hi @andrewfg, in the upcoming weeks I will start working on this topic again. Sorry for the late response. Helping me to modifying the code etc. would be great. I'll get in touch with you once I rebased my branch with current addon branch.

andrewfg commented:
Ok @bruestel cool. Let me know when you have done the rebase.

bruestel commented:
> Ok @bruestel cool. Let me know when you have done the rebase.

Just rebased and updated code base (`mvn spotless:apply` etc.). I will start work on requested changes.

@andrewfg do you experience problems with Server-Sent-Events (SSE)? My installation at home is not not receiving SSE pings after a while.

Hilbrand commented:
Great to see some activity again! I see I had some pending reviews on the other pull request :-( And I see you already integrated the standard OAuth service :tada: To get you started here is some general homework 😉 
1) Remove all references to PaperUI. This UI will be gone in openHAB 3 (including relative links to icons)
2) Drop the logviewer part. As other reviewers already have mentioned. This feature is nice when you provide the binding to users via other means. But once it's integrated it should follow the same standards as every other binding. I won't stop you from contributing to the new UI to improve general logging for all other bindings 💯 
3) Please use the standard jetty http library instead of the okhttp library. It has already become forbidden to add libraries in the lib directory. So that needs to be changed anyway.

One other remark. The last major 2.5. release is planned for 20 september. After that we switch to openHAB 3 development. Given the short timeframe it might be a challenge to get it in 2.5. The changes for openHAB 3.0 development are not big. So changing it won't be a lot of work (I have a post on the forum about this). But it does mean the code won't be compatible with 2.5 anymore. Just something to consider. I hope I didn't discouraged you to much. So let us know if you have any other questions.

bruestel commented:
> Drop the logviewer part. As other reviewers already have mentioned. This feature is nice when you provide the binding to users via other means. But once it's integrated it should follow the same standards as every other binding.

I'll remove the logviewer part. Would it be ok to have some small HTTP request/response buffer? For example, I could show the latest 50 request to the Home Connect API on a special debug-page (servlet based). I would keep the "log-buffer" in-memory to avoid a separate log file. 

>  I won't stop you from contributing to the new UI to improve general logging for all other bindings 100

Will do, I think this is a very important topic as not all users have engineering skills. It should be very easy to filter or export log statements.

> Please use the standard jetty http library instead of the okhttp library. It has already become forbidden to add libraries in the lib directory. So that needs to be changed anyway.

Is it ok to use okhttp in the future? I defined my okhttp dependency in the pom.xml file. In the `lib` directory is just an addon for Server-Sent-Events. I'm thinking about moving to Apache CXF client implementation instead.

> One other remark. The last major 2.5. release is planned for 20 september. After that we switch to openHAB 3 development. Given the short timeframe it might be a challenge to get it in 2.5. The changes for openHAB 3.0 development are not big.

Thanks for the hint. I will switch to 3.

>So changing it won't be a lot of work (I have a post on the forum about this). 
I'm thinking about using the mentioned linux command to create a 2.5.x version of the 3.0.x addon. I would do the steps you've described in the forum post just the other way around (maybe I will create a small script for that). This way I can provide a 2.5.x build to the users, who are already using it.

andrewfg commented:
Hi @bruestel I have some information for you concerning progress that I have made..

- It looks like the OH maintainers have automatically converted your code to the OH3 format
- So I copied the converted code to a new branch `homeconnect-oh3` on my [repository here](https://github.com/andrewfg/openhab-addons/tree/homeconnect-oh3/bundles/org.openhab.binding.homeconnect)
- Then I made a few changes to your code to remove some deprecated/forbidden dependencies
- And I also made a few changes to make it build sweetly (without warnings) in both Eclipse and Maven

My intention is to open a PR from my repository branch to OH3 `main` so that we can get your code adopted asap into OH3. If so, then we would close your PR (#6794) against v2.5.x and open a new one on v3.0.0 `main` from my `homeconnect-oh3` branch instead. => Is that Ok with you?

bruestel commented:
> Hi @bruestel I have some information for you concerning progress that I have made..
> 
> * It looks like the OH maintainers have automatically converted your code to the OH3 format
> * So I copied the converted code to a new branch `homeconnect-oh3` on [my repository here](https://github.com/andrewfg/openhab-addons/tree/homeconnect-oh3/bundles/org.openhab.binding.homeconnect)
> * Then I made a few changes to your code to remove some deprecated/forbidden dependencies
> * And I also made a few changes to make it build sweetly (without warnings) in both Eclipse and Maven
> 
> My intention is to open a PR from my repository branch to OH3 `main` so that we can get your code adopted asap into OH3. If so, then we would close your PR (#6794) against v2.5.x and open a new one on v3.0.0 `main` from my `homeconnect-oh3` branch instead. => Is that Ok with you?

Hi @andrewfg,

can you please have a look at: https://github.com/bruestel/org.openhab.binding.homeconnect/tree/2.5.x-next
Currently I'm heavily refactoring the binding:

- get rid of `/lib/` dependency
- remove own logging framework
- add monitor to SSE connections
- code cleanup and simplification
- applied review feedback
- refactor servlet

andrewfg commented:
> Hi @andrewfg,
>
>can you please have a look at: https://github.com/bruestel/org.openhab.binding.homeconnect/tree/2.5.x-next
>Currently I'm heavily refactoring the binding:

Ah Ok @bruestel then I won't do anything more on my side :)

bruestel commented:
I've created a new PR https://github.com/openhab/openhab-addons/pull/9187
